File: tools/Local Server/controllers/evalTask.js
// controllers/evalTask.js
const { VM } = require('vm2');
const logger = require('../utils/logger');

function handleEvalTask(code) {
    if (!code) {
        return JSON.stringify({ success: false, error: 'No code provided for eval' });
    }

    try {
        const vm = new VM({
            timeout: 1000,
            sandbox: {},
        });
        let result = vm.run(code);
        return JSON.stringify({ success: true, result: result });
    } catch (err) {
        logger.error('Error executing eval task: ' + err.message);
        return JSON.stringify({ success: false, error: err.message });
    }
}

module.exports = handleEvalTask;

File: launcher/src/Tasks/EvalTask.cs
using System;
using System.Threading;
using Microsoft.ClearScript;
using Microsoft.ClearScript.V8;
using Newtonsoft.Json.Linq;

namespace CF7Launcher.Tasks
{
    /// <summary>
    /// ClearScript V8 替代 vm2 沙箱。
    /// 每次 eval 创建新引擎实例，1 秒超时。
    /// 响应：{success:true, result:<value>} 或 {success:false, error:"..."}
    /// </summary>
    public static class EvalTask
    {
        public static string Handle(JObject message)
        {
            string code = message.Value<string>("payload");
            if (string.IsNullOrEmpty(code))
                return Error("No code provided for eval");

            V8ScriptEngine engine = null;
            CancellationTokenSource cts = null;
            try
            {
                engine = new V8ScriptEngine();
                cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(1000));
                cts.Token.Register(delegate { engine.Interrupt(); });

                object result = engine.Evaluate(code);

                JObject resp = new JObject();
                resp["success"] = true;
                resp["result"] = result != null ? JToken.FromObject(result) : JValue.CreateNull();
                return resp.ToString(Newtonsoft.Json.Formatting.None);
            }
            catch (ScriptInterruptedException)
            {
                return Error("execution timed out");
            }
            catch (ScriptEngineException ex)
            {
                return Error(ex.Message);
            }
            catch (Exception ex)
            {
                return Error(ex.Message);
            }
            finally
            {
                if (cts != null) cts.Dispose();
                if (engine != null) engine.Dispose();
            }
        }

        private static string Error(string msg)
        {
            JObject resp = new JObject();
            resp["success"] = false;
            resp["error"] = msg;
            return resp.ToString(Newtonsoft.Json.Formatting.None);
        }
    }
}
